### ISSUE TITLE
Coordinate porting to Windows

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#5430]
*UPDATE 2023-04-24:*

*All the main platform features for Windows are finished! 🚀*

*There are still some smaller stories pending. Project board: https://github.com/orgs/crystal-lang/projects/11/views/5*

*You can support the ongoing development financially: [Windows suppport](https://opencollective.com/crystal-lang/projects/windows-support) project on Open Collective.*

*UPDATE 2021-11-18:*

*By now, the compiler works pretty well on windows and most of the standard library has been ported as well. Besides a couple other smaller missing pieces, the major gap is in concurrency-related features held back by the missing event loop implementation.*

-----

With #5339 being merged, the first step of bringing Crystal to Windows has been accomplished!
With current master branch it is possible to cross-compile and run simple programs (alá *Hello, World!*) on native Windows.

Obviously we're still far away from porting the compiler or entire standard library to windows. This issue is supposed to coordinate and keep track of ongoing efforts to add support for more and more of the crystal standard library.

# TODO

The primary focus should be on the core library (somewhat odered by priority):

* [x]  File descriptors #5339
* [x] GC #5339
* [x] Errno #5339 
* [x] Basic Exceptions #5339 
* [x] Regex #5339
* [x] Random #5539 
* [x]  Time #5422 
* [x]  Raising #6419 
* [x]  File #5623
* [x]  Dir #5623 
* [x] ENV #6333
* [x] Fibers #7995
* [x] Event loop #12149
* [x]  Signal #7339
* [x]  Mutex  #12213
* [x]  Sockets #10696
* [x]  Process
* [x]  Backtraces #11461
* [x]  Iconv (#11480)
* [x]  System #6972
* [x]  Threads (#11647)
* [x] Compiler #9032

And of course a CI infrastructure once specs are running on windows.

# Proceedings

The general course of action is to extract platform-specific implementations and then add the implementation for win32 platform.

A lot of work has already been done in #3582 but it needs to be updated and chopped into smaller pieces of changes.

The first goal should probably be to get specs running on windows to be able to verify the windows implementations. Essential for this are as far as I can see file, dir, exceptions (backtrace!), process and signal.

If you want to help with that effort, or want to "take" a piece of that work to work on, please comment below.

# Useful References 

* Wiki page [*Porting to Windows*](https://github.com/crystal-lang/crystal/wiki/Porting-to-Windows)
* #26 *Support Windows as a first class platform (note: not WSL or emulation)*
* #3582 *[WIP] Windows* by @lbguilherme 
* #5339 *"hello world" windows port* by @RX14
* #4832 *Win: define Win32 API functions* by @txe
* #5448 *Remove c/winapi and rename some windows types* (some pieces from removed c/winapi can be salvaged for implementations)

*Please feel free to add and edit this post as appropriate (core team).*

**Comment by RX14** [comment]
@straight-shoota that's a good point. We need to figure out the basics of getting a blank `spec` suite working on windows (we should do this first without unwind support, just the `raise`==`exit` we have now). Then we need to create some kind of `win32_spec.cr` which requires only the parts of `spec/std` that work.

**Comment by RX14** [comment]
Another area of priority is looking at https://github.com/crystal-lang/crystal/commit/c168035853fe8ee9b06a22e5890414e5efa61776 and working out ways of reducing that stubbing. For example: get iconv working so we can remove the stubbing in `src/io/encoding.cr`. The *modified* parts of the corelib are actually I think worth focus much more than the *missing* parts (for now).

**Comment by straight-shoota** [comment]
A very basic working spec shouldn't be too hard I think. I've tried it earlier to figure out what might be needed for that. It seems to be missing mostly some basic file and dir methods. After disabling all feature options I even got a simple spec to compile, allthough it resulted a LLVM error at codegen.

**Comment by TsundereBug** [comment]
I'll take Fibers, I'll be able to start work on them after January 2nd.

**Comment by RX14** [comment]
@straight-shoota looks like my `spec` PR has turned into a `Dir` PR first.

**Comment by RX14** [comment]
Also of importance: deciding what to do about windows's UTF-16 use. There should definitely be a really quick and easy way of going between String -> UTF-16 `WCHAR*` and the reverse. We should also have a policy of no `A` functions, only `W` functions.

**Comment by RX14** [comment]
Good news: spec compiles and works.

Bad news: it only works sometimes.

From this spec failure, looks like we have a codegen or ABI bug somewhere:
![](https://my.mixtape.moe/wmtqpi.png)

**Comment by RX14** [comment]
Here's my branch porting `spec`: https://github.com/RX14/crystal/commits/feature/windows-spec

I'll turn "Remove c/winbase" and "Add Crystal::System::Dir" into PRs.

**Comment by RX14** [comment]
Another idea before I forget it:

Consolidate all stubbing out into one huge `windows_stubs.cr` file so that we can keep track of them all.

**Comment by straight-shoota** [comment]
You mean to remove winapi.cr, not winbase.
And this commit should also rename DWord.

**Comment by RX14** [comment]
@straight-shoota yep, thanks, commit message is wrong. Shouldn't be working at 1am...

**Comment by RX14** [comment]
Made those PRs: #5447 and #5448.

**Comment by straight-shoota** [comment]
Looks great!

I've deliberatly added a note at the end of the OP for that purpose ;)

**Comment by felipetesc** [comment]
Great news! Thank you for this.

**Comment by drhuffman12** [comment]
In https://github.com/crystal-lang/crystal/wiki/Porting-to-Windows, should:

`Ongoing efforts will be tracked and coordinated in #`

... be:

`Ongoing efforts will be tracked and coordinated in #5430`?

**Comment by straight-shoota** [comment]
@drhuffman12 thanks, fixed!

**Comment by drhuffman12** [comment]
@straight-shoota , yw

**Comment by TsundereBug** [comment]
Due to time constraints and last minute events I won't be able to work on Fibers as expected :(. If someone else wants to claim them, that would be amazing.

**Comment by TsundereBug** [comment]
Due to time constraints and last minute events I won't be able to work on Fibers as expected :(. If someone else wants to claim them, that would be amazing.

**Comment by RX14** [comment]
If you can get openssh server working on WSL (hint: port `22` is taken, and you need to edit `sshd_conf` to set `UsePrivilegeSeparation no`), you can use the following script to make the windows testing process seamless:

```sh
#!/bin/sh
windows=192.168.122.237

bin/crystal build --progress --cross-compile --target x86_64-unknown-windows-msvc -o .build/cross "${@}"
scp .build/cross.o $windows:/mnt/c/crystal/cross.o

cat <<EOF | ssh $windows bash
export PATH="$PATH:/mnt/c/Windows/System32"
cmd.exe /S /c "C:\Program^ Files^ ^(x86^)\Microsoft^ Visual^ Studio\2017\BuildTools\VC\Auxiliary\Build\vcvarsall.bat x64 && cd \crystal && cl cross.o /Fecross pcre.lib gc.lib libcmt.lib"

cd /mnt/c/crystal

echo
./cross.exe
EOF

```

This script assumes you have `gc.lib` and `pcre.lib` in `C:\crystal` and that you're using the visual studio 2017 build tools. MSVC 2015 can surely just change the path to `vcvarsall.bat`.

**Comment by RX14** [comment]
A script to automatically generate a compilable manifest of the entire spec suite, with broken tests commented out with a vague reason:

```sh
cd spec
for spec in ./std/**/*_spec.cr; do
  (cd .. && crystal-windows spec/$spec) >/dev/null 2>/dev/null && echo "require \"$spec\"" ||
    (../bin/crystal build --no-codegen --target x86_64--windows-msvc $spec >/dev/null 2>/dev/null && echo "# require \"$spec\" (failed to run)" || echo "# require \"$spec\" (failed to compile)")
done
```

**Comment by straight-shoota** [comment]
The script posted by @RX14 is intended to be run on an external system which connects to WSL through SSH to upload the object file and run the linker through Windows interoperability feature.
If you want to run the compiler locally on WSL, you don't need to go through SSH.
I have put together a `crystal-windows` wrapper for this: https://gist.github.com/straight-shoota/e4393c9e3f0febb7dc290470f92f5c0d

**Comment by txe** [comment]
As I see, there is no dir.cr in crystal/system/win32, so I'm going to make it.
Is there any idea what to do with UTF-16?

**Comment by txe** [comment]
About UTF-16: I suppose we need just a couple of functions to encode\decode utf-16 without any len, find and others functions and can keep utf-16 string as array of wchar.
FYI, Go has [this code](https://github.com/golang/go/blob/master/src/unicode/utf16/utf16.go) for that and Nim [this](https://github.com/oskca/nsciter/blob/master/widestr.nim)

**Comment by asterite** [comment]
@txe Please coordinate with @RX14 , as far as I know he was porting File to Windows (not sure about Dir). For UTF16 we now have this: https://github.com/crystal-lang/crystal/pull/5541 . But check the PR's comments, we might need to change it a bit to include a zero at the end when encoding to Slice(UInt16).

**Comment by RX14** [comment]
@txe i'm working on `File` and `Dir` right now, they're both mostly complete but not updated in my branch. `UTF-16` is also mostly there in my branch.

I have a lot of work not in `master`, perhaps comitting by branch would help you have more of a base to work on?

**Comment by j8r** [comment]
@RX14 Yes the best is to create a dedicated branch, with a "WIP" pull request to master, even if it's a bit *trashy*. Everyone will know you work on it - this can avoid duplicate work and help others.

**Comment by asterite** [comment]
In fact, since none of this will "matter" until we can bootstrap the compiler in Windows, I would prefer small PRs that do some small functionality, even if incomplete. Then we can merge it (there's no risk of breaking anything) and everyone can already play with that and continue improving it. I prefer this instead of working on a branch.

**Comment by txe** [comment]
Good, I'm glad to hear that there is so much progress.
Then I'll take a look at porting Fiber, it shouldn't be hard, because it already worked in @lbguilherme's branch.

**Comment by RX14** [comment]
@txe Feel free to ignore this suggestion - but actually I think the most important port is stackunwinding and exception handling on windows. You seem to have this work largely done on your old windows branch, it'd be fantastic if you could PR it to master.

**Comment by txe** [comment]
OK, let see what I can do with exceptions.

**Comment by RX14** [comment]
OK, i've completed my `File` port, and `file_spec.cr` now compiles and runs on windows. The results are at [`feature/windows-spec`](https://github.com/RX14/crystal/tree/feature/windows-spec). I've split 3 new PRs out from that branch: #5579, #5580, and #5581.

**Comment by txe** [comment]
What can I do next after exceptions?

**Comment by RX14** [comment]
@txe probably fiber, your original plan would be fairly easy.

I think a lot of things are stuck on me porting spec/File/Dir right now. Though my exams are over so I should be able to get that PRed soon.

**Comment by RX14** [comment]
Just PRed #5622 and #5623. The latter ports `File` and `Dir` to win32.

I changed my "windows" branch to just be called "windows", so if you want to follow the latest work on the windows port it should now be https://github.com/RX14/crystal/tree/windows.

**Comment by ShriekBob** [comment]
Hey.  So, if I'm interested in getting involved here, do we have a setup guide?

**Comment by bew** [comment]
Yes @ShriekBob, you can find setup informations on the wiki: https://github.com/crystal-lang/crystal/wiki/Porting-to-Windows and in this thread.

**Comment by straight-shoota** [comment]
If you encounter any issues, feel free to ask on Gitter/IRC https://gitter.im/crystal-lang/crystal

**Comment by RX14** [comment]
Currently halfway through porting `Process`.

**Comment by nof1000** [comment]
So is there any news on this?

**Comment by RX14** [comment]
Just a few notes to people building pcre and gc:

for bdwgc:
1. clone bdwgc and libatomic_opts repos (use tag v7.6.4)
2. Download https://gist.github.com/ynkdir/688e62f419e5374347bf as ntwin32.mak
3. `nmake -f NT_X64_STATIC_THREADS_MAKEFILE`
4. use `gc.lib`

for pcre:
1. download pcre-8.41
2. `mkdir build && cd build`
3. `cmake .. -G "Visual Studio 15 2017 Win64" -DBUILD_SHARED_LIBS=OFF -DPCRE_SUPPORT_UTF=ON -DPCRE_SUPPORT_JIT=ON -DPCRE_STATIC_RUNTIME=ON`
4. use `release/pcre.lib`
4. `cmake --build . --config release`

**Comment by RX14** [comment]
And a new `crystal-windows` script for use with a windows-native ssh server:

```sh
#!/bin/sh
set -euo pipefail

windows=192.168.122.76

bin/crystal build --progress --cross-compile --target x86_64-unknown-windows-msvc -o .build/cross "${@}"
scp .build/cross.o $windows:'C:\Crystal\cross.obj'

cat <<'EOF' | ssh $windows cmd /S /C "C:\Program^ Files^ ^(x86^)\Microsoft^ Visual^ Studio\2017\BuildTools\VC\Auxiliary\Build\vcvarsall.bat x64 && powershell -Command -"
cd C:\Crystal
del cross.exe

cl cross.obj /Fecross pcre.lib gc.lib advapi32.lib libcmt.lib # C:\data\programming\crystal-lang\crystal\windows\src\llvm\ext\llvm_ext.obj "$(llvm-config --libs --system-libs --ldflags --link-static)"

echo ""
./cross.exe --verbose
EOF
```

**Comment by RX14** [comment]
And for building LLVM:
```
cmake -Thost=x64 -G "Visual Studio 15 2017 Win64" .. -DCMAKE_INSTALL_PREFIX="C:\Program Files\llvm\" -DLLVM_TARGETS_TO_BUILD="X86" -DLLVM_PARALLEL_LINK_JOBS=1 -DLLVM_BUILD_LLVM_DYLIB=ON -DLLVM_USE_CRT_RELEASE=MT
cmake --build . --config release
```

**Comment by straight-shoota** [comment]
@RX14 Are step 4 and 5 for libpcre are in wrong order?

I updated the instructions on the [wiki page](https://github.com/crystal-lang/crystal/wiki/Porting-to-Windows).

**Comment by RX14** [comment]
@straight-shoota yep, thanks

**Comment by RX14** [comment]
@straight-shoota instructions are wrong: should be saved as `ntwin32.mak`

**Comment by rdp** [comment]
Just curious, why target msvc vs. mingw-w64?

**Comment by RX14** [comment]
@rdp far less dependencies and it's not that much more difficult.

**Comment by rdp** [comment]
mingw-w64 is just way easier for cross compiling, but I suppose support can be added for it "later" as it were :)
Also should sockets be added to the list?

**Comment by RX14** [comment]
@rdp cross-compiling is already simple

**Comment by sam0x17** [comment]
msvc is also pretty much as "first class" as you can get on windows, so targeting it is definitely a good idea. This is why big tools like node-gyp use msvc for the most part.

**Comment by straight-shoota** [comment]
> Also should sockets be added to the list?

done.

**Comment by svnpenn** [comment]
I agree with @rdp - anyone who does serious work on Windows knows that MSVC is a few orders of magnitude more bloated than mingw64 - you can get a working GCC with about 26 MB worth of downloads.

**Comment by straight-shoota** [comment]
There's no point in arguing mingw64 would be easier when we've already got msvc working quite nice. And for now the focus should be on actually porting the stdlib to win32 API instead of supporting another compiler tool. Eventually, we'll be supporting both like we have multiple backends for other platforms as well. But right now I don't see any benefit from switching to mingw64.

**Comment by konovod** [comment]
Mingw allows to compile windows programs without using windows or wine at all, using only linux tools and linux compiler. This can be useful in some situations until Windows support is self-hosted, e.g. for people without WSL. As for performance - afaik there is no significant difference - there is no "extra layer" like in cygwin, so both mingw and msvc are "native". Many (most?) opensource tools and libraries support both toolchains.
That said, of course having Windows support is much better then switching the toolchains, and as I remember msvc was chosen because there was some problem with Exceptions handling in mingw.

**Comment by RX14** [comment]
Mingw comes with the win32 API, which is what we're coding against. It's the same API, so all the current work is applicable to both. I wouldn't be surprised if mingw is in fact already supported simply by copying the libc folder to the right triple.

**Comment by svnpenn** [comment]
@RX14 that a dangerous assumption. PHP went the same route Crystal is going, and
now its going to take thousands of lines of code to retrofit mingw64 support.
I know, because I started work on it last year:

- http://bugs.php.net/bug.php?id=75412
- http://github.com/php/php-src/pull/2905

**Comment by RX14** [comment]
@svnpenn correct me if i'm wrong, but that just seems to be minor header definition incompatibilities. Crystal doesn't use or parse header files so we don't have the same problem.

**Comment by svnpenn** [comment]
@RX14 while Crystal is written in Crystal, it does appear to use C at the lower
level:

https://github.com/crystal-lang/crystal/blob/7e8b219aafe921e85bd6344d73ad67a96cd495c9/src/io.cr#L1-L5

which means that it is vulnerable to the same problems as PHP. The commit that
I posted for PHP would be the first of probably 100. Essentially every single
`require` would need to be looked at across the entire Crystal codebase, as:

1. while they have significant overlap - Visual Studio has a different set of
headers from Mingw64

2. again, while they have significant overlap - Visual Studio has a different
set of `typedef`s from Mingw64

I get that the important step is implementing Windows, full stop, but if you
neglect Mingw64 long enough, eventually it wont be possible to add the support
at all.

**Comment by RX14** [comment]
@svnpenn that's referring to `src/lib_c`, crystal's bindings to libc (written in crystal). There's no headers or header parsing involved. As long as it provides all the functions crystal uses, with the same names, and the same arguments, it should be trivial to port.

**Comment by RX14** [comment]
If mingw is "`sizeof()` compatible" with msvc (not even the calling convention or harder parts of the ABI), then the port becomes a single `cp` command.

**Comment by straight-shoota** [comment]
This issue's purpose is to coordinate efforts, not arguing about how to proceed. Please keep this issue clear of such detailed discussions. If someone feels to talk about mingw64 support, open an issue for it.
Thanks.

**Comment by svnpenn** [comment]
https://github.com/crystal-lang/crystal/issues/6170

**Comment by straight-shoota** [comment]
I'll work on porting the missing pieces of `ENV`. Should be fairly straightforward.

**Comment by Neltherion** [comment]
It's always great to hear new progress on Windows... Thanks...

**Comment by Qwerp-Derp** [comment]
Raising should be updated to #6419, shouldn't it? Also, what's the progress on #6333 @straight-shoota?

**Comment by straight-shoota** [comment]
Todo list is updated.

ENV is good to go and waiting for review & second approval.

**Comment by thiagomajesk** [comment]
I'm curious, does this Issue pace represent future Windows support/ investment by the core team or is just a rough path for specific stuff related to Crystal compiler? (No offense, just plain curiosity)

**Comment by RX14** [comment]
@thiagomajesk i'm not quite sure what you mean? We intend to support windows equally as any other platform going into the future

**Comment by thiagomajesk** [comment]
@RX14 I'm relieved to 'hear' this. I'd hate to see Crystal going the path Ruby went, creating a completely closed environment culture (regarding OS support). I was just wondering if this port was getting enough attention as other *nix systems.

**Comment by sam0x17** [comment]
@thiagomajesk a big problem in the ruby ecosystem is a lot of the FFI-based gems (gems that use local compilation of C/C++ to do things) do not support compilation on windows, and therefore are broken on windows. This has the potential to be a problem on windows, but I think the difference is since crystal is a compiled language, there are going to be more people focused on making the cross platform experience better because they want to use crystal to build cross platform desktop apps. This didn't make as much sense with ruby because no one wants to distribute an interpreter-based desktop app, so no one built the desktop app tooling.

**Comment by thiagomajesk** [comment]
@sam0x17 I couldn't agree more. Besides that, I think that the core team has a very important job setting the pace for the community regarding this matter, like discouraging OS specific implementations in favor of cross-platform ones. I'm really anxious to see how things will end up for the community. Thanks for sharing your thoughts.

**Comment by straight-shoota** [comment]
I think we've come quite a good way on the path to Windows support. We're about halfway through in the TODO list. And the already ported stdlib features are probably more complex then the ones still pending. The next major milestone is `Fiber`, which also includes porting `Thread` and `libevent`. A first step in that direction is #6955.

Work on porting the compiler could also begin soon, once the Path API has been merged (#5635).

As an additional note, travis-ci has announced it now [provides CI runners on Windows](https://blog.travis-ci.com/2018-10-11-windows-early-release), which could be a great way to add automatic tests for the compiler and stdlib specs.

**Comment by bew** [comment]
> libevent won't be used on windows

Why? iirc Windows is a supported plateform for libevent

**Comment by asterite** [comment]
@RX14 Pipes can be implemented in pure crystal, there's no need to create OS pipes. Go's [io.Pipe](https://golang.org/pkg/io/#Pipe) works like that.

That said, I guess that pipe isn't very useful to share data when forking... but then again, you can't really fork in Go, so that's not a problem there (and it wouldn't be a problem in Crystal if we stopped supporting `fork`).

**Comment by RX14** [comment]
@asterite you need to be able to spawn processes and send IO to them... We need pipes.

**Comment by RX14** [comment]
And even if pipes weren't a problem, libevent would *still* be using select on windows which has terrible performance. IOCP is the only way to go.

**Comment by asterite** [comment]
Ah, makes sense. I didn't know `libevent` for Windows just supported `select`.

**Comment by RX14** [comment]
BTW I already pretty much ported the compiler in a branch, I just havent had time to work on the branch recently. Porting the compiler depends on porting `Process`.

**Comment by j8r** [comment]
`libuv` seems to support pipes in its [design](http://docs.libuv.org/en/v1.x/design.html) - I haven't the knowledge to confirm this.
I think the best would be to reuse an existing library, unless we have the resources and the knowledge to implement a better, custom solution.

**Comment by RX14** [comment]
I seem to recall there were other reasons crystal didn't use libev/libuv, but not what they were. Either way, they can't be used.

**Comment by RX14** [comment]
Oh yes, the problem is that libuv's event loop isn't thread safe.

**Comment by straight-shoota** [comment]
Please, lets discuss the details of porting fiber in a dedicated issue.

**Comment by markrjr** [comment]
For System is it just cpu_count and hostname that need to be ported?

**Comment by markrjr** [comment]
I have a working implementation of [system for Win32](https://github.com/markrjr/crystal/commit/aa43722cfcb007a564cb48c6d7448005791a9767). I'm not super familiar with Crystal's style guide and I'm not sure if I'm correctly handling the unsafe code in hostname. Any pointers (no pun intended) are welcome.

**Comment by r00ster91** [comment]
Just run `crystal tool format` to follow Crystal's style guide.

**Comment by straight-shoota** [comment]
@markrjr Looks great! Why don't you just open a PR and we can discuss details there?

**Comment by rishavs** [comment]
> I'm also interested in helping get some CI stuff setup for Windows, we could likely use Azure Pipelines.

I actually work in the AzDev Pipelines team. Let me know if there's anything I can help with.

**Comment by rishavs** [comment]
Any updates on the porting? Christmas is coming near <wink wink> <nudge nudge>

**Comment by straight-shoota** [comment]
Unfortunately not much.

* #6972 has been added for porting `System`. It currently waits for implementing review comments.
* #6955 needs some help for implementing the fiber switch on win32 in assembly.
* #5635 is blocking progress with porting the compiler. But I hope this will be ready to merge soon.

**Comment by RX14** [comment]
I'm busy with university currently, but christmas break is coming up soon where I'll likely resume work.

**Comment by Neltherion** [comment]
Thank you all for the hard work. I'm really excited for the day I'm able to port my current Ruby scripts to Crystal on Windows.

**Comment by wdhwg001** [comment]
I think it's time to update the to-do list, #6972 is now merged and is for System instead of Signal.

**Comment by AxDSan** [comment]
I'm so excited to finally see Crystal in windows 💃 Keep on guys. I wish I could help in something

**Comment by maltevoos** [comment]
How are we going to port `Signal` to Windows? From what I've read, `Signal`s aren't supported at all on Windows. Please correct me if I'm wrong.

**Comment by RX14** [comment]
signals are supported: https://docs.microsoft.com/en-us/previous-versions/xdkz3x12(v%3Dvs.140)

**Comment by maltevoos** [comment]
I already looked at this page, and I think these signals are only there because they are part of the C standard library. From [wikipedia](https://en.wikipedia.org/wiki/C_signal_handling):

> The C standard defines only 6 signals. They are all defined in signal.h header (csignal header in C++):
> 
> SIGABRT - "abort", abnormal termination.
> SIGFPE - floating point exception.
> SIGILL - "illegal", invalid instruction.
> SIGINT - "interrupt", interactive attention request sent to the program.
> SIGSEGV - "segmentation violation", invalid memory access.
> SIGTERM - "terminate", termination request sent to the program.
> 
> Additional signals may be specified in the signal.h header by the implementation. For example, Unix and Unix-like operating systems (such as Linux) define more than 15 additional signals; see Unix signal.

And these 6 signals are the only ones "supported" by Windows.

There is already an [issue](https://github.com/nodejs/node/issues/12378) about this topic on the Node.js repo.
sam-github commented there:

> [...] Windows as an O/S has absolutely no signal support. [...]
> 
> If SIGTERM was supported on Windows, you could listen for it and receive it... but you cannot. You can listen..., but you'll never receive it.

And I guess this is exactly the case. You have a function for listening to signals (`signal()`), but there is nothing like `kill()` for sending signals on Windows.

**Comment by rdp** [comment]
IIRC Ruby supports it by translating it (or "faking it"), ex: SIGTERM they
send a Terminate call or what have you...

**Comment by maltevoos** [comment]
I think "faking" signals is not that easy...  For example `SIGTERM`  would politely ask a process to terminate on UNIX. On Windows, you only have `TerminateProcess`, but this would stop the process immediately, like `SIGKILL` on UNIX. The only `SIGTERM` replacement I found would be running taskkill.exe with the PID, which is a rather inelegant solution. My approach would be not to "fake" signals on Windows since this could be misleading and potentially produce different behaviour. I would suggest to find replacements where Crystal itself uses signals (`Process` for example).

**Comment by maltevoos** [comment]
@rdp Probably you are right. I already found some replacements for signals one might want to send:

- SIGTERM: `SendMessage(WM_CLOSE)` <- this is also what `taskkill.exe` uses; works for windowed processes
- SIGKILL: `TerminateProcess()` <- used by `taskkill.exe /f`; works for windowed and console processes
- SIGINT: `GenerateConsoleCtrlEvent` <- works for console processes, simulates Ctrl + C

**Comment by bew** [comment]
This issue is for coordination around windows support, and keeps track of where we are.
I think the discussion around signals on windows would be better in a separate issue (and maybe move the comments there? Not sure it's possible)

**Comment by maltevoos** [comment]
@bew Done. #7339 
I'll keep my comments here so it's clear what this conversation was about. Or should I delete them?

**Comment by hanyuone** [comment]
Should the current build process on Windows be simpler? Some of the steps in the "Porting to Windows" page on the Wiki, such as downloading LLVM and GC, could be automated via shell script, and the `cl` command is really unwieldy if you use libraries. For example, when compiling an example for my library, `hedron`, on Windows, I have to use this:

```
cl control_gallery.o /Fecontrol_gallery _rc2bin\resources.res gc.lib pcre.lib libui.lib user32.lib kernel32.lib gdi32.lib comctl32.lib uxtheme.lib msimg32.lib comdlg32.lib d2d1.lib dwrite.lib ole32.lib oleaut32.lib oleacc.lib windowscodecs.lib msvcrt.lib legacy_stdio_definitions.lib advapi32.lib
```

It'd be nice if the compilation process on Windows were at least simplified - that way, more people could test for problems in the new platform.

**Comment by straight-shoota** [comment]
The wiki page already provides a link to an [installation script](https://gist.github.com/RX14/c6ca69bfbe1f21d30d1e0b5ed3c15199) in the very first step. It should build/install all required libraries. The last section *How to Build Libs* just explains how to build manually.

For compiling Crystal binaries, the cross compilation script should be good enough for most simple use cases. Currently, you can't do much more on windows anyway until we get fiber and proper IO implemented. So I don't expect many people would benefit from experimenting on windows in  the current state.

But IIRC @RX14 said he isn't too far away from porting the compiler to be self-hosting on windows. Going native would obviously simplify the build process and remove the need to cross compile.

**Comment by sam0x17** [comment]
I came across this and thought it might have tidbits useful for the team trying to add windows fibers support: https://nullprogram.com/blog/2019/03/28/

@straight-shoota @RX14 @asterite

**Comment by rishavs** [comment]
Hi guys. Any updates on the overall porting?

**Comment by RX14** [comment]
I've been super busy with uni this year unfortunately, if people want quick progress on windows I suggest they make sure I'm busy with crystal over the summer by sponsoring me.

**Comment by rishavs** [comment]
Hey Chris
How does one go about it?

**Comment by Sija** [comment]
@RX14 perhaps new GH feature would come handy here, see https://help.github.com/en/articles/receiving-sponsorships-as-a-sponsored-developer

**Comment by Pizzacus** [comment]
Woah :O The feature hasn't been here for 24 hours yet and people are already begging for sponsors

**Comment by RX14** [comment]
> @RX14 perhaps new GH feature would come handy here, see https://help.github.com/en/articles/receiving-sponsorships-as-a-sponsored-developer

Well, that's good timing for github to announce that right after I posted.

I'll probably have time to set that up after my exam on Tuesday. If people want to sponsor me with another method I suggest they contact me through the email on my profile.

**Comment by danieldaeschle** [comment]
I'm waiting for it <3
Please finish it. I don't know if i could help. I have no experience with building a compiler.

**Comment by wdhwg001** [comment]
Should we update the Fibers section, changing the related PR from #6955 to #7995 ?

**Comment by mrakobeze** [comment]
I may not know something, but, huh, is it such a necessary to use MSVC? Could not it be easier to use MinGW/MSYS/Cygwin to build Crystal on windows?

**Comment by asterite** [comment]
@mrakobeze It's an easier step for us Crystal developers but it means you need those tools as a developer which is harder for a user. I also think MSVC might be faster than using MinGW/others.

**Comment by straight-shoota** [comment]
@mrakobeze This was already asked and answered in https://github.com/crystal-lang/crystal/issues/5430#issuecomment-387780310 and following

**Comment by Edregol** [comment]
I'm not well informed regarding this topic. But as far as I know using MSVC bring some licensing conditions with it and the only reason using MSVC is to have the linker for the windows platform. As far as I know LLVM brings a complete toolchain for windows targets including a linker. Is there a reason not to pick LLVM? Or am I on the wrong track here?

**Comment by straight-shoota** [comment]
It might eventually be an option to use only the LLVM toolchain, but we not only need a linker but also a number of system libraries and they're provided by MSVC. 

Rust had this idea as as well, but it seems they still haven't gotten anywhere either. It's a huge task and I think we're better off taking the short way to get windows support working first.

**Comment by rdp** [comment]
regarding MinGW see https://github.com/crystal-lang/crystal/issues/6170

**Comment by mrakobeze** [comment]
@asterite as of my expirience, MinGW compiled projects are a bit faster on my machine than MSVC ones. About difficulties to end-developer using MinGW/MSYS/etc, for example ruby needs MSYS env to compile gems native exts, and, as far as I know almost nobody sees this as a problem. And, finally, after MSVC support comes, will there be MInGW/MSYS/etc support? I think I could contribute to porting Crystal for MSYS, but, unfortunately, not MSVC as I don't use it. Moreover, I haven't MSVC on my machine, MSYS only.

**Comment by mrakobeze** [comment]
And finally, support of LLVM Clang on Windows - is it available yet?

**Comment by straight-shoota** [comment]
@mrakobeze Support for MinGW is very welcome! We'd very much appreciate if you wan to contribute to that. There is even a separate issue to track that: #6170
I suppose it's probably going to to be easier, to not run two tracks parallel but wait until MSVC support is at least somewhat stable.

Regarding LLVM on  windows, see my previous comment: https://github.com/crystal-lang/crystal/issues/5430#issuecomment-519410752

**Comment by cfsamson** [comment]
I just wanted to bring up the natural question of what's next? 

Now that basic fiber support is unblocked there are no more tracking issues for the other bullet points. I could possibly contribute with some work on `sockets` depending weather we're talking basic blocking socket support or or if we're thinking of non-blocking, overlapped sockets meant to be used in an IOCP event loop.

**Comment by straight-shoota** [comment]
@cfsamson Basic fiber support is working, but we're still missing the entire event loop, which is a requirement for any IO work. We need at least the first three steps from #6957  to commence porting sockets. The event loop is probably going to be quite a challenge to work out a platform-indepemdemt design and implement. 
We'll need to start a dedicated discussion about that.  Help is of course always welcome.
Other features like backtraces could probably be worked on right now, but they're not critical (although it would certainly help with the ongoing efforts on windows).

**Comment by cfsamson** [comment]
Ok, so I guess #6957 is the best issue to follow going forward then. I've been looking into cross platform event loops covering epoll/kqueue/IOCP for a different project and a safe _and_ performant implementation with an ergonomic api is not a trivial task, but we got to start somewhere.

**Comment by rishavs** [comment]
Now that #8112 is merged, I suppose #6957 and #7995 are next.
Also, should someone start a thread on a platform independent implementation of event loop?

**Comment by wdhwg001** [comment]
Now #7995 has been merged, I suppose it's time to update the to-do list.

Also, maybe we should add another item `Events` to the list?

**Comment by AxDSan** [comment]
9 more implementations left!!! SO CLOSE!

**Comment by RX14** [comment]
Sorry to break the bad news, but we're not close to close, and my resources to work on this have basically vanished. Each one of those remaining tasks is hard, and many API issues still are yet to be solved. Don't expect this any time soon.

**Comment by AxDSan** [comment]
Dang it, anyways thanks for keepin' it real, and letting us know, I thought this was gonna be a breeze :(

**Comment by rubyFeedback** [comment]
People on reddit etc... often reference this issue, which is now ~2 years old.

I would like to recommend to the crystal team to eventually think about closing the issue, BUT continue with a NEW one, including updated stuff, in particular what is missing - since evidently things improved quite a lot in the last 2 years too.

It's difficult for non-crystal centric users to keep track of what is going on in this regard.

**Comment by QuickFixGuy** [comment]
Could most of those issues be solved through the use of Libuv? Is there any disadvantages to using it?

**Comment by straight-shoota** [comment]
This issue is supposed to track and coordinate efforts for porting to win32. For this purpose, it doesn't matter how old it is. The progress tracker in the first post is updated continuously.
Please avoid detail and meta discussions.

**Comment by rdp** [comment]
for discussion on the eventloop (libuv et al) see #6957 (maybe it could be added to the checkbox at the top?)

**Comment by cup** [comment]
I am willing to donate $800 upon completion of this issue, thank you!

**Comment by wontruefree** [comment]
I donate to Crystal generally but is there a way to donate to help this issue move along?

**Comment by oprypin** [comment]
For the purpose of coordination,
I'm currently porting `Process` and `Signal`.

**Comment by oprypin** [comment]
Oh nevermind, guess @jan-zajic is farther along in that anyway.
https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+author%3Ajan-zajic+process

**Comment by RX14** [comment]
@oprypin there hasn't been movement in those PRs for some time, and I'd really love to get them merged. Perhaps you could contact @jan-zajic and work out if they need to be picked up

**Comment by oprypin** [comment]
There has indeed been [a reply](https://github.com/crystal-lang/crystal/pull/8642#issuecomment-610837311), so I picked up `Process` after all.

Tracking this task in https://github.com/crystal-lang/crystal/issues/9032

**Comment by oprypin** [comment]
So, now there's a minimal port of `Process` for Windows (turned out mostly based on @RX14's earlier work, though there's also significant inspiration from @jan-zajic's work), and the Crystal compiler can compile itself on Windows. See https://github.com/crystal-lang/crystal/issues/9032 for details.

I also reworked the [Porting to Windows](https://github.com/crystal-lang/crystal/wiki/Porting-to-Windows) wiki page.

There's still some work I need to get merged, including [advanced features for `Process`](https://github.com/crystal-lang/crystal/pull/8518) by @jan-zajic. It'll be [in my branch](https://github.com/oprypin/crystal/compare/winapi) for now.

**Comment by asterite** [comment]
@oprypin You are AWESOME

**Comment by wontruefree** [comment]
I was having trouble with the old wiki page.  Thanks for the update!!

**Comment by wdhwg001** [comment]
@ oprypin you are a legend! Does that mean we can finally have `crystal_i386.exe` and `crystal_x86_64.exe` compilers in the next release before 1.0 if it gets merged?

(A month ago we were planning to release 1.0 without Windows support... You've changed the world.)

**Comment by oprypin** [comment]
This is not Windows support, just a small but particularly juicy part of it. See https://github.com/crystal-lang/crystal/issues/5430#issuecomment-548759783, that hasn't really changed.

But I suppose yes, distributing some .exe as part of the release should be considered, to simplify a few first steps of the initial setup.

**Comment by oprypin** [comment]
So, Windows support is in a state where you can try it out quite easily and build programs natively, without WSL or cross-compilation. But very important parts of the functionality are still broken or missing -- most notably, networking of any kind.

For the curious, there's [**a quick start guide**](https://pryp.in/blog/28/running-crystal-natively-on-windows-building-videogame-examples.html).

**Comment by rishavs** [comment]
Maybe its time to look at the packaging then. A simple way to install and setup the environment would be great to get people try it out and share feedback.

While a full-on GUI based installer may still be too much at this point in time, but how about adding a scoop installer?

**Comment by oprypin** [comment]
Packaging is https://github.com/crystal-lang/distribution-scripts/issues/64

**Comment by kubo** [comment]
In https://github.com/crystal-lang/crystal/pull/9257#issuecomment-639447695:
> It's intended that we migrate from the "half-posix" file API to the full windows API. I used the unix-like wrappers only because of the ease of storing a fd instead of a `HANDLE*`, but this all needs refactoring to ideally use handles everywhere.

May I create a pull request to use `HANDLE` directly instead of fd in `IO::FileDescriptor`? I think it is necessary to invoke file I/O in the event loop because IOCP requires file handles with `FILE_FLAG_OVERLAPPED`.
It passes cross-compilation of `spec/std_spec.cr` on linux. I have not run it on Windows. I bet there are some flaws.

**Comment by rubyFeedback** [comment]
Is there an update to this? Any showstoppers?

**Comment by oprypin** [comment]
* The event loop is very difficult to get right (or at all).
    * Nobody has touched that topic in probably a year by now.
* Almost all prior contributors to Windows aren't currently motivated.

That said, I'm happily making a video game in Crystal and it's running just fine on Windows. Distribution story is very good too; so easy to get a self-contained executable (no problems with static linking like elsewhere).

**Comment by incognitorobito** [comment]
I was thinking about taking a wack at signals, which while not the biggest piece would certainly be one more piece of the puzzle. The discussion started off really great but then just died so I'm not sure where we stand. The work @malte-v did seems like it would work extremely well.

**Comment by maxwellb** [comment]
Long time follower, first time replier. I'm not sure what kind of bandwidth I have, but are there any areas I could chime in on?

Also, if there are any resolutions from earlier, where reviewing alternate approaches for a sanity check would help, that could be fun, too. Thanks!

**Comment by rishavs** [comment]
You guys can probably take a crack at the event loop which seems to holding up everything else. https://github.com/crystal-lang/crystal/issues/6957

**Comment by rubyFeedback** [comment]
If I understand the situation then it may be that some crystal scripts would work on native
windows right now (if we ignore WSL, because on WSL things should be fine) whereas
others don't. Is there some way to get some overview as to what works and what does
not? Perhaps this may help get more people involved to tackle whatever isn't yet 
working. Even some support could be fine for many scripts, e. g. if you just have something
that populates local files with data, that probably already works on windows? I am asking
because I may be looking at a new programming language finally (got a bit fatigued 
recently from ruby). I am eyeing python or crystal as a possible transition language.

**Comment by konovod** [comment]
This question better fits to [forum](https://forum.crystal-lang.org/), but basically you can see status at the top of page. Fibers and networking doesn't work on Windows, (almost) everything else does.

**Comment by chances** [comment]
Please update the Sockets task item in the issue description to point to: https://github.com/crystal-lang/crystal/issues/10696

**Comment by HertzDevil** [comment]
For encodings we should probably link to GNU libiconv on Windows. Maybe we can eventually do it on selected Unix platforms too (for example Android adheres to POSIX but its `iconv_*` functions only support ASCII and UTF-*).

**Comment by frnco** [comment]
> # TODO
> The primary focus should be on the core library (somewhat odered by priority):
> 
> * [x]   File descriptors ["hello world" windows port #5339](https://github.com/crystal-lang/crystal/pull/5339)
> * [x]  GC ["hello world" windows port #5339](https://github.com/crystal-lang/crystal/pull/5339)
> * [x]  Errno ["hello world" windows port #5339](https://github.com/crystal-lang/crystal/pull/5339)
> * [x]  Basic Exceptions ["hello world" windows port #5339](https://github.com/crystal-lang/crystal/pull/5339)
> * [x]  Regex ["hello world" windows port #5339](https://github.com/crystal-lang/crystal/pull/5339)
> * [x]  Random [Implement Crystal::System::Random for win32 #5539](https://github.com/crystal-lang/crystal/pull/5539)
> * [x]   Time [Implement Crystal::System::Time for win32 #5422](https://github.com/crystal-lang/crystal/pull/5422)
> * [x]   Raising [Support exception handling on windows #6419](https://github.com/crystal-lang/crystal/pull/6419)
> * [x]   File [Implement File and Dir for win32 #5623](https://github.com/crystal-lang/crystal/pull/5623)
> * [x]   Dir [Implement File and Dir for win32 #5623](https://github.com/crystal-lang/crystal/pull/5623)
> * [x]  ENV [Extract platform-specifics from ENV to Crystal::System::Env and implement for win32 #6333](https://github.com/crystal-lang/crystal/pull/6333)
> * [x]  Fibers [Implement fibers for win32 #7995](https://github.com/crystal-lang/crystal/pull/7995)
> * [ ]  Event loop [Port concurrency features to Windows #6957](https://github.com/crystal-lang/crystal/issues/6957)
> * [ ]   Signal [Port signals to Windows #7339](https://github.com/crystal-lang/crystal/issues/7339)
> * [ ]   Mutex
> * [x]   Sockets [Porting `Socket` to win32 #10696](https://github.com/crystal-lang/crystal/issues/10696)
> * [x]   Process
> * [ ]   Backtraces
> * [ ]   Iconv
> * [x]   System [Implement System for Win32 #6972](https://github.com/crystal-lang/crystal/pull/6972)
> * [ ]   Threads
> * [x]  Compiler [Bootstrapping compiler on Windows (tracking bug) #9032](https://github.com/crystal-lang/crystal/issues/9032)

This list could probably use a bit more information about the state of each feature. In the beginning of the post it says the compiler is still far away, but then it's marked as done in the list, and between "Nothing" and "Completed" we have a gradient of states, so how well should one expect it to work? The list being binary like this means things get a check either when they are working minimally well, or when they are working well enough to be relied on by devs. I hope it's the second but the list could use some explanation on what's the actual criteria for the checks, and maybe also some details/gotchas wherever makes sense.

If, as I hope, the checks on the list mean everything that doesn't deal with concurrency is actually working in windows, then that is actually quite amazing news, and I'll start using crystal for multi-platform stuff (And maybe even contributing to this effort).


@oprypin mentioned lack of manpower (devpower?) to work on this, but also said he's making a game and it's running fine on windows:
> * The event loop is very difficult to get right (or at all).
>   
>   * Nobody has touched that topic in probably a year by now.
> * Almost all prior contributors to Windows aren't currently motivated.
> 
> That said, I'm happily making a video game in Crystal and it's running just fine on Windows. Distribution story is very good too; so easy to get a self-contained executable (no problems with static linking like elsewhere).

Perhaps a good way to draw more people is making it clearer what can and can't be done in windows. I, for one, really like programming in crystal, but usually I can't share the stuff I write in Crystal with friends and family because Windows (I'm on Arch Linux BTW). If I'm right about this list and concurrency is the last missing ling, there's a ton of stuff I can already do, and if I'm doing a lot of Windows stuff in Crystal it will be a lot easier to help with this. I'm talking about me but I'm sure it is true for many people. I'm using mint-lang and loving it, and it doesn't compile in Windows yet because of Crystal, I'm pretty sure more people want to get Mint working in Windows, and if concurrency is the only obstacle left we should really make an effort to spread that information.



Also, to @rubyFeedback, about your post:
> If I understand the situation then it may be that some crystal scripts would work on native windows right now (if we ignore WSL, because on WSL things should be fine) whereas others don't. Is there some way to get some overview as to what works and what does not? Perhaps this may help get more people involved to tackle whatever isn't yet working. Even some support could be fine for many scripts, e. g. if you just have something that populates local files with data, that probably already works on windows? I am asking because I may be looking at a new programming language finally (got a bit fatigued recently from ruby). I am eyeing python or crystal as a possible transition language.

From one Rubyist to another, it's been around 5 years since I started to really work on learning new languages (not that I hadn't learned many languages already, by that time besides Ruby I had also worked with C, Pascal, PHP, Javascript, some Python, ASP.NET, Java, all the usual suspects, basically, plus many others which I didn't use as much so i wouldn't be able to remember them all, though I do remember mIRC scripting since it was what I started with). But Ruby was just so much better, easier to work with, and let's face it, for an interpreted language Ruby has a ridiculously good performance, gems are also way lighter than node modules or rust crates, so no matter which language I tried it just felt inferior to Ruby. Then about 5 years ago I just got fed up with how everything I did involved Ruby, I really wanted to get a better understanding of the functional programming paradigm and I really wanted to try writing a bunch of stuff in compiled languages to get a better understanding of the pros and cons. I was also interested in trying out a strongly-typed language, but that was not the priority and only recently I came to realize when strongly-typed languages are actually helpful. Anyways, at that time I started to seriously try a lot of different languages, and I think some of what I learned the hard way may be useful for you, so you have a bit more information about what you should and/or shouldn't expect from some languages, and what are some of the reasons for learning or not learning them. I was just gonna include those reasons here but midway it became obvious to me that it would get fairly long and wouldn't be as easy to share with others who may find it useful, plus I realized there are some friends who I also wanna share this with, so I decided to make it more detailed and extensive and created a Gist for that, maybe you will find it helpful: https://gist.github.com/frnco/3f4384d438e1bc0b64c6c56fe00e7fdf

**Comment by straight-shoota** [comment]
Hey @frnco, this issue is mostly used as an internal tracker of the development status. The original post is almost four years old, so I think it's expectable that the introductory motivation isn't up to date anymore (otherwise there would be no progress). I guess it wouldn't hurt to put an update on the current, greatly improved status there. But as I said, this is not intended as public documentation of supported windows features.

Regarding the task list, every checked item means the respective feature is fully implemented to a state that we seem appropriate for shipping, maybe with some edge cases left out (but that applies to all other code as well).

> everything that doesn't deal with concurrency is actually working in windows

Yes, that's mostly it. There are a couple other smaller missing pieces, but by now the major gap is in concurrency-related features held back by the missing event loop implementation. However, I would expect to have some improvements there soon.

Apart from that, there's really a lot you can do already on Windows. With https://github.com/marketplace/actions/install-crystal it's easy to run specs in CI.

But: Windows support is not yet ready. We are not publishing distribution packages for windows and installing is a bit more effort.

**Comment by rishavs** [comment]
> We are not publishing distribution packages for windows 

This is the part which I am unable to understand and I hope you can share more on this rationale. If we enable installation on Windows, at least we enable folks to use other use cases of Crystal, even if the event loop is missing.

I would love to use Crystal for gamedev and console apps. But just because the networking part isn't there I can't use Crystal for such use cases. At least not in a user friendly way.

IMO, adding distribution/installation scripts for Crystal on Windows alongwith compiler warnings around the missing networking (and any other feature) would be a great short term solution.

The work on networking has been ongoing for years and it will likely take a long time still. Why block a lot of genuine use cases for Crystal till then?

**Comment by neatorobito** [comment]
Agree with @rishavs, here's a proof of concept that builds on the porting page and the work of @oprypin on his blog.

Install the VS build tools using the [instructions from the porting page](https://github.com/neatorobito/scoop-crystal), then using scoop package manager:
`scoop bucket add crystal-preview https://github.com/neatorobito/scoop-crystal`
`scoop install crystal`

Restart the terminal and voila:

![image](https://user-images.githubusercontent.com/3013405/142776010-cb0e0e7d-06be-4128-97bb-165c03a0afec.png)

If we publish the build artifacts from the Windows CI for the next release, then I can update the links, add the correct hashes, and get everything buttoned up.

An install workflow this simple combined with open communication/warnings around unsupported/under construction parts of the language would really drive more attention towards the porting efforts and the language as a whole.

**Comment by straight-shoota** [comment]
@neatorobito I don't know scoop. But that looks neat. I don't think you need any official distribution package for this. It seems grabbing the build artifacts from GitHub actions is enough for your setup. I assume that's what the zip file contains.
I don't think it makes sense to pin that to a specific nightly release, so you just need to point the download URL should point to the latest nightly build. That's available through the GitHub API. You can look at https://github.com/crystal-lang/install-crystal/blob/8ce664227e1b9494770e1ca29515538f2fe29797/index.js#L410-L444 how it can be done. Or maybe https://nightly.link/ could be an option.

Regarding an official distribution package, I believe we still need to figure out some practical details.

**Comment by neatorobito** [comment]
Yup, the zip file just has the artifacts. I'm having some trouble downloading the artifact directly from GitHub, however. If I hit the API using a powershell version of script you linked I get this:

```
https://api.github.com/repos/crystal-lang/crystal/actions/artifacts/116800750/zip
Invoke-WebRequest : {"message":"You must have the actions scope to
download artifacts.","documentation_url":"https://docs.github.com/rest/refe
rence/actions#download-an-artifact"}
```

Seems like we either need a public token with the read actions scope, or to add nightly.link to the crystal repo and I can use that instead. I vote for the latter.

**Comment by frnco** [comment]
> But: Windows support is not yet ready. We are not publishing distribution packages for windows and installing is a bit more effort.

I did try to compile stuff for Windows and that became pretty obvious. Still, seems like it could be pretty beneficial to have more people trying to use Crystal in Windows ASAP, more people using it means more people testing, finding bugs, and hopefully contributing to the port.

**Comment by oprypin** [comment]
Regarding a direct download:

Trying to download the CI artifact from GitHub is absolutely not workable, and indeed it would need a token, and publishing (hardcoding) a token (for the purpose of putting it into an installer script or otherwise) just isn't allowed by GitHub's terms.

https://nightly.link/crystal-lang/crystal/workflows/win/master/crystal.zip works indeed as a "permanent" URL but that's 100% a hack when used as part of automation. What you download can be different every time and particular downloads literally just expire after some time.

So I fully agree that a versioned and stable download is very much needed. Maybe especially for the next release (looking to be an important one for Windows). And I had already asked about that here: https://github.com/crystal-lang/distribution-scripts/issues/64 - that has some good discussion but I think it didn't have a good reason to be fully stalled.

Maybe that particular discussion of just getting a stable download URL for the artifacts should be revived.

---

> The work on networking has been ongoing for years and it will likely take a long time still. Why block a lot of genuine use cases for Crystal till then?

It's not necessarily blocked on any particular core feature. It's blocked on itself! That is, to publish a distributable version of Crystal, one would need to *make* it distributable -- make installing and running it make sense on Windows, rather than requiring a series of workarounds. And it's very difficult to organize the final steps there.

I'm guessing that the reluctancy has been just that among official Crystal downloads everything is always readily runnable, without hacks, and also has stable relative paths within the archive, etc. Whereas publishing a Windows package right now would necessarily mean eventually publishing a completely differently put together package.

---

> I would love to use Crystal for gamedev and console apps. But just because the networking part isn't there I can't use Crystal for such use cases.

And you can indeed use it for those! And indeed it'll be the networking, not the installation process blocking you there. So is this even an argument for the purposes of this discussion?

---

> Still, seems like it could be pretty beneficial to have more people trying to use Crystal in Windows ASAP, more people using it means more people testing, finding bugs, and hopefully contributing to the port.

I'm sorry I'll say something really not nice, but this really compelled me to respond.

I'm guessing that people who aren't willing to manually copy some files and set the environment also wouldn't be willing to seriously contribute. There have already been several breakthrough moments for Windows support, but in the end it's almost always been long-time contributors that have been able to productively contribute.

And as for finding bugs, well, for the moment the loose ends are quite obvious.

But perhaps at a certain stage of Windows support this sentiment will actually become true. And maybe that stage is nearing, with all the ongoing work by @straight-shoota and @HertzDevil.

**Comment by HertzDevil** [comment]
With the recent support for `CRYSTAL_LIBRARY_PATH` we should be able to package "portable" Windows archives in the same directory structure as the ones for Linux and macOS.

The MSVC build tools are an unavoidable dependency for now, but if we package our releases with [`vswhere.exe`](https://github.com/microsoft/vswhere) we might be able to skip the part about launching the MSVC developer prompt prior to running Crystal. (The compiler itself needs to invoke `vswhere.exe` and assemble the correct path for `cl.exe` and so on.) Some other languages prefer to install their own MSYS(2) instances for their Windows installers and I strongly prefer not to do that.

**Comment by neatorobito** [comment]
@HertzDevil I've come up with a solution where the scoop package depends on vswhere, then I have a [shim](https://github.com/neatorobito/scoop-crystal/blob/main/scripts/crystal.ps1) that loads the relevant dev environment variables and calls the crystal compiler.

**Comment by oprypin** [comment]
I support the idea of making the compiler itself call `vswhere` if `cl` is not in PATH. In the interest of avoiding wrapper scripts, which I believe can lead to weird side effects on Windows.

And yes, we have the ability to hardcode `CRYSTAL_LIBRARY_PATH` into the compiler.

Having that, perhaps we could indeed just rely on `scoop` to provide MSVC and `vswhere` in addition to Crystal. Probably wouldn't would want to actually *ship* those components.

---

@neatorobito I was just wondering, in that repo, what *is* `vs_2022_cpp_build_tools` and where does it come from, if there are 0 references to it in all of GitHub and Google?

---

And umm too bad that we already started reusing this issue for the specific question of packaging. But oh well.

**Comment by neatorobito** [comment]
@oprypin It's a package that installs the VS Build Tools for the Crystal compiler, specifically the `Microsoft.VisualStudio.Workload.VCTools` and `Microsoft.VisualStudio.Component.VC.ATL` workloads.

I made it because there isn't a generic VS Build Tools package in any of the other scoop repos since VS has so many different workloads.

>Having that, perhaps we could indeed just rely on scoop to provide MSVC and vswhere in addition to Crystal. Probably wouldn't would want to actually ship those components.

This is what I've setup in my [repo](https://github.com/neatorobito/scoop-crystal), please give it a try if you get a chance.

**Comment by oprypin** [comment]
@neatorobito That doesn't answer what I actually want to know.

> It's a package that installs the VS Build Tools for the Crystal compiler,

*Where* is that package? How to browse its code? What is the reason, according to whatever Scoop does, that that package can be found, despite me being unable to find it on Google search?

**Comment by neatorobito** [comment]
The json manifest declares that itself along with whatever scoop grabs and executes is a "package".

In this case it's just `vs_BuildTools.exe` from `https://aka.ms/vs/17/release/vs_BuildTools.exe` and then runs it with the arguments.

**Comment by straight-shoota** [comment]
Please, let's continue this discussion in a focussed issue. This is a meta issue for tracking progress of individual components.

**Comment by frnco** [comment]
> Please, let's continue this discussion in a focussed issue. This is a meta issue for tracking progress of individual components.

I agree that we should split the discussion, and in that interest started a new thread about the compilation process with windows as a target, which is still quite rough: https://github.com/crystal-lang/crystal/issues/11486

But it's also worth mentioning that we lack a reasonable way to create issues about windows, should we classify it as "Language Improvement Discussion" or what? How should we keep track of everything? Should we link all the discussions to this thread? Adding an option for "Windows-related" issues when starting an issue could make things a lot easier.

**Comment by oprypin** [comment]
@straight-shoota Where would you suggest to discuss automated installation, such as the proposed one with Scoop? I wish there was a neat way to move comments starting from https://github.com/crystal-lang/crystal/issues/5430#issuecomment-974878009 to a new issue

**Comment by straight-shoota** [comment]
I think #11486 is the place to continue this discussion. For specific distribution channels such as scoop, maybe a dedicated issue would be better.

**Comment by rubyFeedback** [comment]
Would it be possible to update the table showing how much work is still to be done?

1.5.0 recently was released and improved the windows support.

I myself use Linux (and ruby too) but I'd love to use crystal on windows "out of the
box", for other people (in particular elderly people who struggle with computers).
Crystal being faster than ruby is a HUGE advantage here; I could use Java but
Java annoys me way more than ruby does, yet ruby is really too slow on windows,
we have to admit this.

Ideally a progress bar could be shown in percentage rather than 19 / 22, and
perhaps a colour code, e. g. green, red, orange, yellow - a bit like how the
WineHQ project tracks the support of different games.

**Comment by HertzDevil** [comment]
If every sub-item's corresponding issue/PR is closed then we could consider it done, but of course it is not as simple as that. The largest missing pieces in the standard library are:

* Non-blocking file I/O (there was an attempt [here](https://github.com/crystal-lang/crystal/issues/6957#issuecomment-752801567), and I wonder if a Win32 thread pool is truly necessary here)
* The general sentiment in #7339 is to implement replacement APIs for `Signal` rather than port it to MSVC
* `System::User` and `System::Group`, or whether that POSIX API even makes sense on Windows (#7738)
* [Unix domain sockets](https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/)?
* `-Dpreview_mt`

And in the compiler itself:

* The playground builds but hangs on the first run. Is `Process#wait` meant to be non-blocking?
* I can confirm the interpreter can be built with minimal workarounds
* #11575 is preventing us from distributing LLVM libraries in a viable manner and, to a lesser extent, MPIR and Libiconv

**Comment by straight-shoota** [comment]
Some personal comments:

* I don't think non-blocking file IO has a high priority. This seems to be rather uncommon on Windows, actually. That's similar on most POSIX systems btw., where it's probably easier to do, but still often not used by default (File in Crystal is blocking unless you configure it to be non-blocking).
* Non-blocking pipes are certainly important. I think this shouldn't be hard to implement, using the proper Win32 API for that. This just wasn't necessary (or even possible to test) before we got the event loop working.
* Unix domain sockets in Windows are a rather recent feature overall. Adding support for that in Crystal is nice, but IMO not high priority.
* Multithreading support shouldn't be far out I suppose?
* IMO the playground has not much priority. If we can get it to work on Windows, that's great. But the interpreter is likely more important.

So overall, I think there's not much standing in the way of proclaiming Windows as a supported platform.

My must-have list for that is rather short (maybe I'm missing something, though).
* Signals (or equivalent)
* non-blocking pipe
* Confirm concurrent IO operations in practice. The event loop port is very fresh and we should test it with some real work scenarios and stress tests to iron out the edge cases.

**Comment by HertzDevil** [comment]
_Regular_ file I/O could stay blocking, but pipes are `IO::FileDescriptor`s and I don't think there is any non-blocking API specific to pipes that wouldn't generally work on other files.

**Comment by konovod** [comment]
> * I don't think non-blocking file IO has a high priority. This seems to be rather uncommon on Windows, actually.
Just my 2 cents, but it's useful to, say, work with (virtual) COM Port. 
But if general concurrency will work, i think i can go with WinAPI or own dll to emulate non-blocking api.

**Comment by jgaskins** [comment]
> pipes are `IO::FileDescriptor`s

I was actually surprised `IO.pipe` was implemented in terms of file descriptors. I assumed it was basically an `IO::Memory` that used separate read/write cursors until I ran out of file descriptors a while back [with NATS object store](https://github.com/jgaskins/nats/blob/512199fca5fe4603f7f4d57200860e87b61f78b0/src/objects.cr#L175).

Is there any reason not to implement `IO.pipe` in user space so it would work on all platforms?

**Comment by asterite** [comment]
Right, popes in Go are implemented in pure Go, no file descriptors at all. We wanted to do that eventually with Juan but we never got to it.

**Comment by jgaskins** [comment]
Yeah, when I realized, I assumed you used POSIX FDs simply because that was quicker at the time and didn’t intend that to be part of the semantics of the pipe.

Is there an open issue for this? I have some terrible code written for it that mostly works.

**Comment by HertzDevil** [comment]
The most relevant discussion seems to be in #8152.

But if we are passing pipes across processes, possibly ones that aren't written in Crystal, wouldn't that require an OS-backed pipe anyway?

**Comment by straight-shoota** [comment]
Let's continue the discussion about re-implementing pipes in Crystal in #12238

**Comment by 89z** [comment]
Just saw @straight-shoota make this statement on another site:

> Windows support is pretty stable and almost complete by now.

With all due respect, I disagree. I am a Windows developer. Any language implementation that only support Visual Studio, is not complete in my opinion. At a minimum, LLVM, GNU or basically anything but Visual Studio should also be supported:

https://github.com/crystal-lang/crystal/issues/6170

Visual Studio is a giant dependency, and I hope that other developers would agree that its kind of a cop out to say "well it works with Visual Studio, good enough!". Visual Studio takes about 100 times the disk space as other options. Plenty of other languages don't put this burden on Windows developers, including:

1. Rust
2. Go
3. PHP
4. Python
5. Ruby
6. JavaScript
7. C#
8. C++
9. Perl

**Comment by Blacksmoke16** [comment]
I think he means from a stdlib usability point of view. I.e. you can actually use Crystal on windows fairly stably, even if it requires Visual Studio. Having something working and usable is more important to the end user compared to which backend is used when the latter is something that can be added in after the former is working.

See https://github.com/crystal-lang/crystal/issues/5430#issuecomment-395958620.

**Comment by 89z** [comment]
@Blacksmoke16, that comment was four years ago. At that time, it was a valid and reasoned perspective. Now it just comes off as an excuse. If `stdlib` is as "stable and complete" as Crystal developers are now publicly stating, then it seems the time has come to consider adding and/or changing Windows backends.

**Comment by 89z** [comment]
look like the command is set here:

https://github.com/crystal-lang/crystal/blob/b56e786b2f82ac62e0c057680b608842739252c6/src/compiler/crystal/compiler.cr#L371

where `cl` is defined here:

https://github.com/crystal-lang/crystal/blob/b56e786b2f82ac62e0c057680b608842739252c6/src/compiler/crystal/compiler.cr#L340

and `CL` is hardcoded here:

https://github.com/crystal-lang/crystal/blob/b56e786b2f82ac62e0c057680b608842739252c6/src/compiler/crystal/compiler.cr#L23-L24

if Crystal could change `CL` to try an environment variable first, simlar to `CC`, then it should be possible to set that environment variable to `clang-cl`, in order to use LLVM instead of Visual Studio.

**Comment by 89z** [comment]
https://github.com/crystal-lang/crystal/pull/12440

**Comment by mjblack** [comment]
> look like the command is set here:
> 
> https://github.com/crystal-lang/crystal/blob/b56e786b2f82ac62e0c057680b608842739252c6/src/compiler/crystal/compiler.cr#L371
> 
> where `cl` is defined here:
> 
> https://github.com/crystal-lang/crystal/blob/b56e786b2f82ac62e0c057680b608842739252c6/src/compiler/crystal/compiler.cr#L340
> 
> and `CL` is hardcoded here:
> 
> https://github.com/crystal-lang/crystal/blob/b56e786b2f82ac62e0c057680b608842739252c6/src/compiler/crystal/compiler.cr#L23-L24
> 
> if Crystal could change `CL` to try an environment variable first, simlar to `CC`, then it should be possible to set that environment variable to `clang-cl`, in order to use LLVM instead of Visual Studio.

A little further down in the file it is ignoring what `CL` is set to for msvc due to hard coding the executable as a method argument:

https://github.com/crystal-lang/crystal/blob/b56e786b2f82ac62e0c057680b608842739252c6/src/compiler/crystal/compiler.cr#L360

It probably should be updated to:
```
cl = Process.quote_windows(msvc_path.join("bin", "Host#{host_bits}", target_bits, cl).to_s)
```

**Comment by mominshaikhdevs** [comment]
Have been following this thread for soome time. 
(ignoring developing dlls) Is Signal the only remaining major thing at this point ?

**Comment by mdwagner** [comment]
I just ran into the issue where Windows can't find `cl.exe`. I was using `crystal-1.7.1-windows-x86_64-msvc-unsupported.zip`, but will try using Scoop instead.

**Comment by neatorobito** [comment]
The scoop package has the same issue for the moment, still trying to coordinate a fix.

**Comment by mdwagner** [comment]
@neatorobito Ah, you're right. Same issue.

**Comment by neatorobito** [comment]
@mdwagner I've tested and pushed a fix for this in the scoop repo, please give it a try when you get a chance.

**Comment by straight-shoota** [comment]
@mdwagner Can you please open a dedicated issue for this? It's out of scope for this meta issue. Please add details on your system setup (like install location of VS Build Tools).

**Comment by mominshaikhdevs** [comment]
Isn't it high time this + https://github.com/crystal-lang/crystal/issues/26 + https://github.com/crystal-lang/crystal/issues/7339 be closed ?  (since as per [this comment](https://github.com/crystal-lang/crystal/issues/7339#issuecomment-1178327261), https://github.com/crystal-lang/crystal/issues/7339 is completed as https://github.com/crystal-lang/crystal/issues/12224 + https://github.com/crystal-lang/crystal/issues/12284 being completed) and move focus to [other issues](https://github.com/crystal-lang/crystal/labels/platform%3Awindows)?

**Comment by straight-shoota** [comment]
I'm not sure if the signal story is entirely finished. We just recently merged a bunch of changes. If that's all, yes we can close #7339. But we'll need to confirm that.

That would then mean the bulk of the Windows port is completed as per the TODO list in this issue. But there are still a lot of very rough edges, so #26 is still unresolved.

**Comment by zw963** [comment]
Hi, from the TODO list, it's seem like Signal should works on windows, but it build failed on my github action platform is ms-windows 2022.

```
Dependencies are satisfied
Building: translater
[1/13] Parse                             
[1/13] Parse                             
[2/13] Semantic (top level)              
[2/13] Semantic (top level)              
[3/13] Semantic (new)                    
[3/13] Semantic (new)                    
[4/13] Semantic (type declarations)      
[4/13] Semantic (type declarations)      
[5/13] Semantic (abstract def check)     
[5/13] Semantic (abstract def check)     
[6/13] Semantic (restrictions augmenter) 
[6/13] Semantic (restrictions augmenter) 
[7/13] Semantic (ivars initializers)     
[7/13] Semantic (ivars initializers)     
[8/13] Semantic (cvars initializers)     
[8/13] Semantic (cvars initializers)     
[9/13] Semantic (main)                   
Error target translater failed to compile:
Showing last frame. Use --error-trace for full trace.

In lib\selenium\src\selenium\service.cr:92:20
 92 | process.signal(Signal::KILL)

                     ^-----------
Error: undefined constant Signal::KILL
Did you mean 'Signal::ILL'?
```
Please have a look, following is log.

https://github.com/crystal-china/translater/actions/runs/4906504148/jobs/8761004602

**Comment by HertzDevil** [comment]
They should use `Process#terminate`, because `Signal` is not portable, and MSVC never defines `Signal::KILL`. This is by design

**Comment by Neltherion** [comment]
9 Years passed and still no Windows support. This is why Crystal will unfortunately always be a niche language for a small minority.

**Comment by mominshaikhdevs** [comment]
> 9 Years passed and still no Windows support. This is why Crystal will unfortunately always be a niche language for a small minority.

All I have seen here for the last 4 years is, New Issues are piling up everyday. Instead of resolving the old ones. 😂

**Comment by cloewen8** [comment]
> 9 Years passed and still no Windows support. This is why Crystal will unfortunately always be a niche language for a small minority.

Not surprised, checked the funding and it just was not there. I've moved on to other languuages with better marketing. The ecosystem is moving on, without Crystal.

**Comment by Blacksmoke16** [comment]
I mean have any of you actually tried it? Windows support is, and has been for a while, at the point where you can pretty much install via the provided installed on the GH release and use it as you would on any other OS.

There are still some loose ends that need to be figured out, but I don't think it's fair to say "there is no windows support."

**Comment by mjblack** [comment]
> 9 Years passed and still no Windows support. This is why Crystal will unfortunately always be a niche language for a small minority.

You can compile on windows, you can create windows gui applications, you can create directx applications. What do you think is still missing?

**Comment by aolko** [comment]
> > 9 Years passed and still no Windows support. This is why Crystal will unfortunately always be a niche language for a small minority.
> 
> You can compile on windows, you can create windows gui applications, you can create directx applications. What do you think is still missing?

modern web ui libs like electron/tauri?

**Comment by linuxct** [comment]
> > 9 Years passed and still no Windows support. This is why Crystal will unfortunately always be a niche language for a small minority.
> 
> You can compile on windows, you can create windows gui applications, you can create directx applications. What do you think is still missing?

The interactive CLI, actually:
```powershell
PS ..\Users\linuxct> crystal.exe i
Crystal was compiled without interpreter support
```

**Comment by Blacksmoke16** [comment]
> modern web ui libs like electron/tauri/webview? Some libs for AI crowd like Torch, diffusers, transformers and other related dependencies?

There is a difference between the language missing something and the ecosystem missing something. All of these are examples of the latter. Unless you have specific windows specific reason as to why you can't make something like what you mention, then someone just needs to do it. You're not going to get these things in the stdlib. Could always start a discussion on the [forums](https://forum.crystal-lang.org/) to see if you can find some others to help creating/maintaining these kind of libs.

> The interactive CLI, actually:

This is not specific to windows either. The interpreter is still experimental, on all OSs. If you actually want to use it, you'd have to build from source.

**Comment by Blacksmoke16** [comment]
> because windows is not yet a stable, feature complete platform.

> That's why you can't dismiss these concerns with "everything basically works".

I'm just going to repeat my comment from https://github.com/crystal-lang/crystal/issues/5430#issuecomment-2171575519. Do you have any concrete examples of something not working on Windows? If so, please say what it is, or better yet create an issue for it! Otherwise these claims that "there is no windows support" are baseless and are not helping, and seem to only be for trolling purposes...

**Comment by tmssngr** [comment]
If nothing is left on Windows, then better mention that Windows is officially supported now.

**Comment by felipetesc** [comment]
I've been paying attention to this topic for a while now, and I can't let it go unnoticed without thanking you for the effort to make Crystal functional on Windows. Finally, I can use it for general projects. So, thanks.

**Comment by rdp** [comment]
Might be nice to update the description here, it was confusing to me today that it has sections saying it's "missing event loop" and "far from porting the compiler", cheers! :)

**Comment by rdp** [comment]
Might be nice to name the installer exe something different from the portable name, ex: "crystal-1.13.1-windows-x86_64-msvc-unsupported-installer.exe" just to make it even more clear...An idea, cheers!

**Comment by sdogruyol** [comment]
Bump, all TODOs are complete 👍 What's next?

**Comment by Blacksmoke16** [comment]
@sdogruyol Was some discussion on this in https://github.com/crystal-lang/crystal-book/pull/772#issuecomment-2379528748.

**Comment by crysbot** [comment]
This issue has been mentioned on **Crystal Forum**. There might be relevant details there:

https://forum.crystal-lang.org/t/crystal-core-developer-looking-for-work-in-germany-europe/7389/2

**Comment by hamidb80** [comment]
is it done? I see every TODOs has been checked

**Comment by straight-shoota** [comment]
Yes, the major part is finished. But we're not there all the way yet as mentioned in the UPDATE notice at the top of the OP.

I suppose we can probably close this issue though, it has served its purpose.

**Comment by crysbot** [comment]
This issue has been mentioned on **Crystal Forum**. There might be relevant details there:

https://forum.crystal-lang.org/t/user-space-pipe-implementation/8837/24
